src/lib/optimizer/strategy/dependent_group_by_reduction_rule.hpp

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "abstract_rule.hpp"
+#include "expression/abstract_expression.hpp"
+#include "expression/lqp_column_expression.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+class AggregateNode;
+struct TableConstraintDefinition;
+class StoredTableNode;
+
+/**
+ * In SQL, the returned columns of an aggregate are either "group-by columns" or "aggregate columns, e.g., SUM(a)".
+ * As an example, in TPC-H Q10, the aggregate on customer includes both c_custkey as well as c_name (amonst others) in
+ * the group-by columns.
+ * Note how c_name is included as a group-by column, even though the groups are already exhaustively defined by
+ * GROUP BY c_custkey. The reason for that is the functional dependency c_custkey => c_name, or, more practically
+ * speaking, the fact that c_custkey is the primary key of the customer table.
+ * This rule identifies cases where functionally dependent columns are given as group-by columns while their primary
```

Comment:
Yes and no. I believe for Hyrise in its current state, this is correct but in general it could be more precise. A data dependency pedant would probably trip over it.

Smart ass excursus on FDs (you might want to skip reading this): A FD X -> Y, where X (determinant) and Y (dependent) are sets of columns is defined as: if two tuples are equal in X, they will also be equal in Y. Duplicates are possible, FDs do not say anything about uniqueness. Therefore, FDs themselves have nothing to do with primary keys.
There are also UniqueColumnCombinations (UCCs) also called candidate keys. UCCs are defined as: a set of columns K is unique iff there are no two tuples which are equal in K. Primary keys are UCCs. A UCC induces trivial FDs. The UCC columns are the determinant (c_custkey in above's case) and all other columns of the table are possible dependents. While UCCs induce FDs, FDs do not guarantee UCCs.

Now it becomes a bit more interesting again: this optimization works not only in the presence of primary keys but also for simple FDs. FDs are sufficient to remove dependent columns. We could rephrase the comment to make it more precise: "This rule identifies cases where functionally dependent columns are given as group-by columns while the FD's determinant column(s) is a/are also group by column(s)." However, currently we only support unique columns as soft constraints and cannot differentiate between weaker FDs and stronger UCCs, right?

Reply:
I don’t have comment at hand on my mobile, but don’t we solely talk about FDs? Uniqueness is not required and would render the rule useless after most joins anyways.

Reply:
Yes, we only need FDs, but the reader might think that primary keys are necessary or that FDs need primary keys even though that's not the case.

Reply:
Thanks!

Reply:
@Bensk1 : thanks for the clear up. Is the current description ok?

Reply:
yes